File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/Middleware.kt
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */

package software.amazon.smithy.swift.codegen

import software.amazon.smithy.codegen.core.Symbol

abstract class Middleware(private val writer: SwiftWriter, shapeSymbol: Symbol) {
    open val id: String get() = typeName
    open val typeName: String = "${shapeSymbol.name}Middleware"
    open val properties: MutableMap<String, Symbol> = mutableMapOf()

    abstract fun generateInit()

    /**
     * Called after rendering the middleware struct, used for writing extensions
     * (although you can technically put anything here)
     */
    open fun renderExtensions() {}
}

File: Sources/ClientRuntime/Middleware/Middleware.swift
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0.

import class Smithy.Context

public protocol Middleware {
    associatedtype MInput
    associatedtype MOutput

    /// The middleware ID
    var id: String { get }

    func handle<H: Handler>(context: Context,
                            input: MInput,
                            next: H) async throws -> MOutput
    where H.Input == MInput, H.Output == MOutput
}

extension Middleware {

    public func eraseToAnyMiddleware() -> AnyMiddleware<MInput, MOutput> {
        return AnyMiddleware(self)
    }
}
